### PR TITLE
Add multithreaded pattern/fp generator

### PR DESCRIPTION
**Pull request by bp-kelley**
This is a simple addition that allows faster creation of libraries.

```
holder = rdSubstructLibrary.CachedSmilesMolHolder()
for smi in pdb_ligands:
      holder.AddSmiles(smi)
    
lib = rdSubstructLibrary.SubstructLibrary(holder)
rdSubstructLibrary.AddPatterns(lib, numThreads=-1)
```

This comes with a more full wrapping of the substructlibrary

### CODE REVIEW

File: Code/GraphMol/SubstructLibrary/substructLibraryTest.cpp [hunk 1]
```diff
@@ -399,6 +401,63 @@ void ringTest() {
...
+
+  boost::shared_ptr<CachedSmilesMolHolder> holder =
+    boost::make_shared<CachedSmilesMolHolder>();
+
+  for(auto s : pdb_ligands) {
+    holder->addSmiles(s);
+  }
+
+  SubstructLibrary ssslib(holder);
+  std::vector<int> num_threads = { 1, 0 };
```
Comment (greglandrum) [inline Code/GraphMol/SubstructLibrary/substructLibraryTest.cpp]: Don't you need an `#ifdef RDK_TEST_MULTITHREADED` somewhere in here to make sure you don't try a multi-threaded run when the code isn't built with multithreading enabled?

Comment (bp-kelley) [reply Code/GraphMol/SubstructLibrary/substructLibraryTest.cpp]: No, in pattern_factory.cpp if only uses one thread when not building in multithreaded mode.

Comment (greglandrum) [reply Code/GraphMol/SubstructLibrary/substructLibraryTest.cpp]: Ah, I see now. So this is then just redundant (but not wrong). Cool

**Comment by bp-kelley** [comment]
@greglandrum If this compiles, I think it's ready.

**Pull request merged**